### PR TITLE
cockpit.format_number() returns empty string for undefined

### PR DESCRIPTION
**Pull request by stefwalter**
If the number input is undefined we these formatting functions return an empty string. This is for formatting displayable output and this is appropriate display behavior in this case, rather than throwing cryptic exceptions.
    
This affects cockpit.format_number(), cockpit.format_bytes(), cockpit.format_bytes_per_sec(), and cockpit.format_bits_per_sec().
    
If callers wish to handle this case differently from the default behavior they should do so before or after calling these methods.

### CODE REVIEW

**Comment by stefwalter** [comment]
Failure seen here:

```
# testRaid (check_storage_mdraid.TestStorage)
#
Page error: TypeError: undefined is not an object (evaluating 'e.toFixed')
http://10.111.113.110:9090/cockpit/$1533cce8c70df10aea804aa7cdc9c667fdb6a7d8/base1/bundle.js 6 Re
http://10.111.113.110:9090/cockpit/$1533cce8c70df10aea804aa7cdc9c667fdb6a7d8/base1/bundle.js 6 W
http://10.111.113.110:9090/cockpit/$1533cce8c70df10aea804aa7cdc9c667fdb6a7d8/base1/bundle.js 6 qe
http://10.111.113.110:9090/cockpit/$1533cce8c70df10aea804aa7cdc9c667fdb6a7d8/storage/bundle.js 30 _
http://10.111.113.110:9090/cockpit/$1533cce8c70df10aea804aa7cdc9c667fdb6a7d8/storage/bundle.js 22 r
 0 map
http://10.111.113.110:9090/cockpit/$1533cce8c70df10aea804aa7cdc9c667fdb6a7d8/storage/bundle.js 22 b
http://10.111.113.110:9090/cockpit/$1533cce8c70df10aea804aa7cdc9c667fdb6a7d8/base1/bundle.js 12 dispatch
http://10.111.113.110:9090/cockpit/$1533cce8c70df10aea804aa7cdc9c667fdb6a7d8/base1/bundle.js 12 handle
http://10.111.113.110:9090/cockpit/$1533cce8c70df10aea804aa7cdc9c667fdb6a7d8/base1/bundle.js 12 trigger
http://10.111.113.110:9090/cockpit/$1533cce8c70df10aea804aa7cdc9c667fdb6a7d8/base1/bundle.js 12 triggerHandler
http://10.111.113.110:9090/cockpit/$1533cce8c70df10aea804aa7cdc9c667fdb6a7d8/storage/bundle.js 14 
http://10.111.113.110:9090/cockpit/$1533cce8c70df10aea804aa7cdc9c667fdb6a7d8/base1/bundle.js 12 dispatch
http://10.111.113.110:9090/cockpit/$1533cce8c70df10aea804aa7cdc9c667fdb6a7d8/base1/bundle.js 12 handle
http://10.111.113.110:9090/cockpit/$1533cce8c70df10aea804aa7cdc9c667fdb6a7d8/base1/bundle.js 6 s
http://10.111.113.110:9090/cockpit/$1533cce8c70df10aea804aa7cdc9c667fdb6a7d8/base1/bundle.js 6 a
http://10.111.113.110:9090/cockpit/$1533cce8c70df10aea804aa7cdc9c667fdb6a7d8/base1/bundle.js 6 w
http://10.111.113.110:9090/cockpit/$1533cce8c70df10aea804aa7cdc9c667fdb6a7d8/base1/bundle.js 6 
http://10.111.113.110:9090/cockpit/$1533cce8c70df10aea804aa7cdc9c667fdb6a7d8/base1/bundle.js 6 C
```

**Comment by larsu** [comment]
Why don't we throw a `RangeError` in these cases?

**Comment by stefwalter** [comment]
> Why don't we throw a RangeError in these cases?

Because we're rendering something for display. If someone wants to render undefined for display, I think an empty string is a good way to render it. This allows the default way to display of a number to update as values go from unknown to known. If a caller wants to handle unknown values specifically they still can.

What are the benefits of throwing RangeError?

We discussed this some on IRC, but happy for further discussion.

File: doc/guide/cockpit-util.xml [hunk 1]
```diff
@@ -41,22 +41,62 @@ array = cockpit.format_bytes(number, [factor, separate])
...
 </programlisting>
     <para>Format <code>number</code> of bytes into a displayable speed <code>string</code>.</para>
+
+    <para>If specifying 1000 or 1024 is specified as a <code>factor</code> then an appropriate suffix
+      will be chosen. By default the <code>factor</code> is 1024.  You can pass a string suffix as a
+      <code>factor</code> in which case the resulting number will be formatted with the same suffix.</para>
+
+    <para>If the <code>number</code> is less than the <code>factor</code> or an unknown factor
+      was passed in, then the formatted number is returned without a suffix. If <code>separate</code>
+      is true, returns an array of <code>[formatted_number, suffix]</code> unless no suffix is returned.</para>
```
Comment (dperpeet) [inline doc/guide/cockpit-util.xml]: This paragraph is not entirely clear on what will happen if "no suffix is returned". A reader will ask if that is connected to the first part of the paragraph or not defined.

Comment (stefwalter) [inline doc/guide/cockpit-util.xml]: Fixed.

**Comment by larsu** [comment]
> What are the benefits of throwing RangeError?

Catching errors early. I think the chance that there's an error in client code is pretty high when a variable that is supposed to contain a number is `undefined`. Getting an exception thrown is much more visible and easier to test for than an empty string somewhere on the page.

Logic for how to display a number when it's value is not yet available should be much higher than in this utility function.

I don't want to block this, though. Especially not if there was already a discussion about it. Just adding my opinion :)

**Comment by dperpeet** [comment]
I like the change, especially if we pass along data from another source. For special handling, I think the caller should ensure to check values.

File: doc/guide/cockpit-util.xml [hunk 2]
```diff
@@ -40,23 +40,66 @@ array = cockpit.format_bytes(number, [factor, separate])
 
     <para>If the <code>number</code> is less than the <code>factor</code> or an unknown factor
       was passed in, then the formatted number is returned without a suffix. If <code>separate</code>
-      is true, returns an array of <code>[formatted_number, suffix]</code> unless no suffix is returned.</para>
+      is true, returns an array of <code>[formatted_number, suffix]</code> or
+      <code>[formatted_number]</code> if returned without a suffix.</para>
+
+    <para>If <code>number</code> is <code>null</code> or <code>undefined</code> an empty string or array
+      will be returned.</para>
+  </refsection>
```
Comment (mvollmer) [inline doc/guide/cockpit-util.xml]: I think an empty array is problematic since the existing clients don't handle it.  They will put `result[0]` on the screen, for example, which is "undefined", no?

**Comment by mvollmer** [comment]
What about documenting that it is a bug to pass non-numbers into the functions, but still be nice and not throw an exception?  We could log a warning.

File: pkg/base1/cockpit.js [hunk 3]
```diff
@@ -1379,7 +1387,9 @@ function basic_scope(cockpit, jquery) {
 
         var ret;
 
-        if (suffix)
+        if (string_representation === "")
+            ret = [];
```
Comment (mvollmer) [inline pkg/base1/cockpit.js]: What about skipping this case and just returning `[ "" ]` or ´[ "", suffix ]`?  Otherwise this is a API break.

**Comment by stefwalter** [comment]
So after discussion on IRC, we agreed to have consistent default behavior when passing in null or undefined. If callers need to handle this state in a non-default way, they should do so before calling the format functions.

File: pkg/base1/cockpit.js (hunk 3 above)
Comment (stefwalter) [inline pkg/base1/cockpit.js]: Makes sense. Now returns ```[ "" ]``` here. It makes no sense to return a suffix here, and that is a minor change. Defining this behavior can be considered part of stabilizing this API.

File: doc/guide/cockpit-util.xml [hunk 4]
```diff
@@ -40,23 +40,66 @@ array = cockpit.format_bytes(number, [factor, separate])
...
+  <refsection id="cockpit-format-number">
+    <title>cockpit.format_number()</title>
+<programlisting>
+string = cockpit.format_number(number)
+</programlisting>
+    <para>Formats <code>number</code> into a displayable <code>string</code>. If the number is not
+      an integer, it is rounded to a single decimal place precision. If the number is near zero, but
+      not quite zero it is rounded up or down to a single decimal place.</para>
+
+    <para>If <code>number</code> is <code>null</code> or <code>undefined</code> an empty string or array
```
Comment (dperpeet) [inline doc/guide/cockpit-util.xml]: This isn't true anymore, I think. It can't return an empty array anymore.

Comment (larsu) [inline doc/guide/cockpit-util.xml]: Indeed, thanks for the catch

**Comment by dperpeet** [comment]
I have a fixup for the doc here: https://github.com/dperpeet/cockpit/commit/1fa857df6d5e3f220ed836459611f730c17ce92a

**Comment by stefwalter** [comment]
Squashed dperpeet's patch in.
